controllers/elfmachine_controller_test.go

```diff
@@ -542,6 +562,371 @@ var _ = Describe("ElfMachineReconciler", func() {
+				fake.InitOwnerReferences(ctrlContext, elfCluster, cluster, elfMachine, machine)
+
+				mockVMService.EXPECT().GetCluster(elfCluster.Spec.Cluster).Return(towerCluster, nil)
+				mockVMService.EXPECT().GetVMPlacementGroup(util.GetVMPlacementGroupName(machine)).Return(placementGroup, nil)
+				mockVMService.EXPECT().FindByIDs([]string{*vm2.ID}).Return([]*models.VM{vm2}, nil)
+				mockVMService.EXPECT().AddVMsToPlacementGroup(placementGroup, gomock.Any()).Return(task, nil)
+
+				reconciler := &ElfMachineReconciler{ControllerContext: ctrlContext, NewVMService: mockNewVMService}
+				ok, err := reconciler.reconcilePlacementGroup(machineContext, vm)
+				Expect(ok).To(BeFalse())
+				Expect(err).To(BeZero())
+				Expect(elfMachine.Status.TaskRef).To(Equal(*task.ID))
+				Expect(logBuffer.String()).To(ContainSubstring("Waiting for the VM to be added to the placement group"))
+				expectConditions(elfMachine, []conditionAssertion{{infrav1.VMProvisionedCondition, corev1.ConditionFalse, clusterv1.ConditionSeverityInfo, infrav1.JoiningPlacementGroupReason}})
+			})
+
+			It("should not migrate VM when VM is running and kcp.Spec.Replicas != kcp.Status.UpdatedReplicas", func() {
+				towerCluster := fake.NewTowerCluster()
+				towerCluster.HostNum = util.TowerInt32(2)
+				hostID1 := fake.UUID()
+				hostID2 := fake.UUID()
+				towerCluster.Hosts = []*models.NestedHost{{ID: util.TowerString(hostID1)}, {ID: util.TowerString(hostID2)}}
+				vm := fake.NewTowerVM()
+				status := models.VMStatusRUNNING
+				vm.Status = &status
+				vm.EntityAsyncStatus = nil
+				vm.Host = &models.NestedHost{ID: util.TowerString(hostID2)}
+				elfMachine.Status.VMRef = *vm.LocalID
+				vm2 := fake.NewTowerVM()
+				vm2.Host = &models.NestedHost{ID: util.TowerString(hostID2)}
+				placementGroup := fake.NewVMPlacementGroup([]string{*vm2.ID})
+				kcp.Spec.Replicas = pointer.Int32(2)
+				kcp.Status.UpdatedReplicas = 1
+				ctrlContext := newCtrlContexts(elfCluster, cluster, elfMachine, machine, secret, kcp)
+				machineContext := newMachineContext(ctrlContext, elfCluster, cluster, elfMachine, machine, mockVMService)
+				fake.InitOwnerReferences(ctrlContext, elfCluster, cluster, elfMachine, machine)
+
+				mockVMService.EXPECT().GetCluster(elfCluster.Spec.Cluster).Return(towerCluster, nil)
+				mockVMService.EXPECT().GetVMPlacementGroup(util.GetVMPlacementGroupName(machine)).Return(placementGroup, nil)
+				mockVMService.EXPECT().FindByIDs([]string{*vm2.ID}).Return([]*models.VM{vm2}, nil)
+
+				reconciler := &ElfMachineReconciler{ControllerContext: ctrlContext, NewVMService: mockNewVMService}
+				ok, err := reconciler.reconcilePlacementGroup(machineContext, vm)
+				Expect(ok).To(BeTrue())
+				Expect(err).To(BeZero())
+				Expect(logBuffer.String()).To(ContainSubstring("KCP rolling update in progress, skip migrate VM"))
+			})
+
+			It("should migrate VM when VM is running and host where VM in is not in unused hosts", func() {
```

Comment:
"should migrate VM to another host when the VM is running and the host of VM is not in unused hosts"